pmd-java/src/main/java/net/sourceforge/pmd/lang/java/typeresolution/ClassTypeResolver.java

```diff
@@ -329,6 +412,169 @@ private Class getVariableNameType(Scope scope, String image, Class accessingClas
+     * @return JavaTypeDefinition of the {@code genericType}.
+     */
+    private JavaTypeDefinition getNextTypeDefinition(JavaTypeDefinition context, Type genericType) {
+        if (genericType == null) {
+            return null;
+        }
+
+        if (genericType instanceof Class) { // Raw types take this branch as well
+            return getDefaultUpperBounds(context, (Class) genericType);
+        } else if (genericType instanceof ParameterizedType) {
+
+            ParameterizedType parameterizedType = (ParameterizedType) genericType;
+            JavaTypeDefinitionBuilder typeDef = JavaTypeDefinition.builder((Class) parameterizedType.getRawType());
+
+            // recursively determine each type argument's type def.
+            for (Type type : parameterizedType.getActualTypeArguments()) {
+                typeDef.addTypeArg(getNextTypeDefinition(context, type));
+            }
+
+            return typeDef.build();
+        } else if (genericType instanceof TypeVariable) {
+            int ordinal = getTypeParameterOrdinal(context.getType(), ((TypeVariable) genericType).getName());
+            if (ordinal != -1) {
+                return context.getGenericArgs().get(ordinal);
+            }
+        } else if (genericType instanceof WildcardType) {
+            Type[] wildcardUpperBounds = ((WildcardType) genericType).getUpperBounds();
+            if (wildcardUpperBounds.length != 0) { // upper bound wildcard
+                return getNextTypeDefinition(context, wildcardUpperBounds[0]);
+            } else { // lower bound wildcard
+                return JavaTypeDefinition.build(Object.class);
+            }
+        }
+
+        return null;
+    }
+
+    /**
+     * Returns the ordinal of the type parameter with the name {@code parameterName} in {@code clazz}.
+     *
+     * @param clazz         The Class with the type parameters.
+     * @param parameterName The name of the type parameter.
+     * @return The ordinal of the type parameter.
+     */
+    private int getTypeParameterOrdinal(Class clazz, String parameterName) {
+        TypeVariable[] classTypeParameters = clazz.getTypeParameters();
+
+        for (int index = 0; index < classTypeParameters.length; ++index) {
+            if (classTypeParameters[index].getName().equals(parameterName)) {
+                return index;
+            }
+        }
+
+        return -1;
+    }
+
+    /**
+     * Returns true if the class is generic.
+     *
+     * @param clazz The Class to examine.
+     * @return True if the Class is generic.
+     */
+    private boolean isGeneric(Class clazz) {
+        if (clazz != null) {
+            return clazz.getTypeParameters().length != 0;
+        }
+
+        return false;
+    }
+
+    /**
+     * Contains Class -> JavaTypeDefinitions map for raw Class types. Also helps to avoid infinite recursion.
+     */
+    private Map<Class, JavaTypeDefinition> classToDefaultUpperBounds = new HashMap<>();
```

Comment:
Also, move field declarations to the top, so they are together and easier to find. This relates to #361, eating [our own dogfood](https://pmd.github.io/pmd-5.7.0/pmd-java/rules/java/design.html#FieldDeclarationsShouldBeAtStartOfClass)